src/random.rs

```diff
@@ -0,0 +1,264 @@
+
+unsafe impl PyTypeInfo for PyBitGenerator {
+    const NAME: &'static str = "PyBitGenerator";
+    const MODULE: Option<&'static str> = Some("numpy.random");
+
+    fn type_object_raw<'py>(py: Python<'py>) -> *mut ffi::PyTypeObject {
+        const CLS: GILOnceCell<Py<PyType>> = GILOnceCell::new();
+        let cls = CLS
+            .get_or_try_init::<_, PyErr>(py, || {
+                Ok(py
+                    .import("numpy.random")?
+                    .getattr("BitGenerator")?
+                    .downcast_into::<PyType>()?
+                    .unbind())
+            })
+            .expect("Failed to get BitGenerator type object")
+            .clone_ref(py)
+            .into_bound(py);
+        cls.as_type_ptr()
+    }
+}
+
+/// Methods for [`PyBitGenerator`].
+pub trait PyBitGeneratorMethods<'py> {
+    /// Acquire a lock on the BitGenerator to allow calling its methods in.
+    fn lock(&self) -> PyResult<PyBitGeneratorGuard<'py>>;
+}
+
+impl<'py> PyBitGeneratorMethods<'py> for Bound<'py, PyBitGenerator> {
+    fn lock(&self) -> PyResult<PyBitGeneratorGuard<'py>> {
+        let capsule = self.getattr("capsule")?.downcast_into::<PyCapsule>()?;
+        let lock = self.getattr("lock")?;
+        if lock.call_method0("locked")?.extract()? {
+            return Err(PyRuntimeError::new_err("BitGenerator is already locked"));
+        }
+        lock.call_method0("acquire")?;
+
+        assert_eq!(capsule.name()?, Some(c"BitGenerator"));
+        let ptr = capsule.pointer() as *mut npy_bitgen;
+        let non_null = match NonNull::new(ptr) {
+            Some(non_null) => non_null,
+            None => {
+                lock.call_method0("release")?;
+                return Err(PyRuntimeError::new_err("Invalid BitGenerator capsule"));
+            }
+        };
+        Ok(PyBitGeneratorGuard {
+            raw_bitgen: non_null,
+            _capsule: capsule.unbind(),
+            lock: lock.unbind(),
+            py: self.py(),
+        })
+    }
+}
+
+impl<'py> TryFrom<&Bound<'py, PyBitGenerator>> for PyBitGeneratorGuard<'py> {
+    type Error = PyErr;
+    fn try_from(value: &Bound<'py, PyBitGenerator>) -> Result<Self, Self::Error> {
+        value.lock()
+    }
+}
+
+/// [`PyBitGenerator`] lock allowing to access its methods without holding the GIL.
+pub struct PyBitGeneratorGuard<'py> {
+    raw_bitgen: NonNull<npy_bitgen>,
+    /// This field makes sure the `raw_bitgen` inside the capsule doesn’t get deallocated.
+    _capsule: Py<PyCapsule>,
+    /// This lock makes sure no other threads try to use the BitGenerator while we do.
+    lock: Py<PyAny>,
+    /// This should be an unsafe field (https://github.com/rust-lang/rust/issues/132922)
+    ///
+    /// SAFETY: only use this in `Drop::drop` (when we are sure the GIL is held).
+    py: Python<'py>,
+}
+
+// SAFETY: we can’t have public APIs that access the Python objects,
+// only the `raw_bitgen` pointer.
+unsafe impl Send for PyBitGeneratorGuard<'_> {}
```

Comment:
This is not sound. `drop` is public, so you if you drop inside of `allow_threads` you're calling into Python without holding the gil.

Reply:
True, it’s possible to drop it inside `allow_threads`. Too bad!

I wish we could get to the lock’s C API, then we could use that instead.